### ISSUE TITLE
Performance against base Michi

### ISSUE THREAD

**Issue by gonmf** [traveller42/michi-go#1]
Great work! Be sure to do a comparison of the two programs later, winrates, playouts per second, ... would be interesting to learn the performance difference of Go.

**Comment by traveller42** [comment]
Test currently running against Michi and Michi-c.  Since each of these are implementing the same algorithm with the same settings, I'm hoping that they are roughly the same strength.

The Go version is a touch faster than the Python version using PyPy in initial testing.

I'll include a more rigorous evaluation once Version 1 is complete.

**Comment by traveller42** [comment]
Against the latest commit (e047812), I have the following results:
michi-py v michi-go (100/100 games)
board size: 13   komi: 7.5

| engine | wins | black | white |
| --- | --- | --- | --- |
| michi-py | 79 79.00% | 41 82.00% | 38 76.00% |
| michi-go | 21 21.00% | 12 24.00% | 9  18.00% |
|  |  | 53 53.00% | 47 47.00% |

Using ./michi-go tsbenchmark:
Tree search with 1400 playouts took 7.025625144s with 8 threads; speed is 24.909 playouts/thread/s

Using ./michi.py tsbenchmark:
Tree search with 1400 playouts took 13.008s with 8 threads; speed is 13.453 playouts/thread/s

I think there may still be an issue with the strength of my translation and I know I can improve the speed.

**Comment by gonmf** [comment]
Am I reading this right, michy-go is (much) weaker but twice as fast?

**Comment by traveller42** [comment]
That is correct.  I still think I have an issue that may be affecting the play.

Just those two alone would imply an ELO difference of 230.

I'm learning much about Python, Golang, and MCTS with this endeavor.  I am quite sure the remaining deficit is due to some misunderstanding on my part.

**Comment by traveller42** [comment]
Using the large patterns, I am now seeing winrates in the 24% range yielding an ELO difference of 200.

This gets much worse when playing without the large pattern files against Michi (also w/o the large patterns).  I haven't found anything obvious, but I suspect something involving the RAVE calculations.  (Mostly because this is the part I least understand.)

**Comment by traveller42** [comment]
I've released version 1.

This is still showing a winrate of 24% against michi-py.

I don't have a good explanation for this difference, but these are the best ideas at this time:
- Some logical error introduced in my translation of the Python
- A difference in the Random Number Generator
- A difference in the calculation of the RAVE value

**Comment by xcombelle** [comment]
I would suspect the Random Number Generator last, as it is not so important in go.

I would like to point that float python are float64 not float32.

A good project to narrow where the weakness could come would be to trace python execution  (with https://docs.python.org/3.5/library/sys.html#sys.settrace) and compare with the result of go functions/

**Comment by traveller42** [comment]
I tend to agree, particularly as I have tried a different RNG with minimal impact.

Thanks for the note regarding float64.  That will actually simplify a few things.

I do intend to eventually dig into the internals and figure out what is different.  (Unless someone else beats me to it.)

**Comment by traveller42** [comment]
I found a translation bug in the move generation code.

I'm running a test now to evaluate the impact.  Initial results show some improvement, but not as much as I was hoping for.

**Comment by traveller42** [comment]
Most recent test run had michi-go with the move generation fix score 10-10 against michi(master).

Most recent test run of tsbenchmark with current master (2a79992):
Tree search with 1400 playouts took 4.626896193s with 8 threads; speed is 37.822 playouts/thread/s

I think there is still some basic performance gains to be had before I get more creative with the code.
